### PR TITLE
[OP] Implementations for prod, nansum and nanprod

### PR DESCRIPTION
**Pull request by alex-weaver**
Added ops and test cases: prod, nansum and nanprod (see issue #3201)

Added test utility function for checking array equality (almost_equal_combo) because reldiff does not work well when all the elements of both arrays are close to zero. For global reductions of prod() the results are often in the order of ground truth = [1e-20], op output = [0]. For this case, reldiff gives a result of 1.0

Instead, almost_equal_combo uses an element-wise threshold in addition to reldiff - if either of them indicate the op output is close to ground truth then the test passes.

Additionally, almost_equal_combo ignores cells in *both* arrays if the corresponding cell in either array is NaN. This allows it to be used to verify the gradients of nansum and nanprod, where what we care about is getting the gradient of non-NaN inputs correct.

### CODE REVIEW

File: python/mxnet/test_utils.py [hunk 1]
```diff
@@ -134,6 +134,29 @@ def assert_almost_equal(a, b, threshold=None):
         raise Exception(msg)
     return rel
 
+def almost_equal_combo(a_src, b_src, threshold=None):
```
Comment (piiswrong) [inline python/mxnet/test_utils.py]: almost_equal_ignore_nan?

Comment (piiswrong) [inline python/mxnet/test_utils.py]: consider follow numpy interface and use atol & rtol
https://docs.scipy.org/doc/numpy-dev/reference/generated/numpy.allclose.html#numpy.allclose

**Comment by alex-weaver** [comment]
Yep that's a much better name for the equality check - I've renamed it and amended the api to include rtol and atol parameters.

**Comment by alex-weaver** [comment]
Just noticed the template specialization makes the build fail (I built with VC++ which is non-compliant in supporting such usage) Will fix and update.

**Comment by alex-weaver** [comment]
That should have resolved it, although for some reason the checks aren't running on commit 35adb28

**Comment by alex-weaver** [comment]
Fixed lint errors

**Comment by alex-weaver** [comment]
(Finally) fixed GPU test failures

**Comment by piiswrong** [comment]
thanks

**Pull request merged**